### PR TITLE
Expressions to convert to and from base64

### PR DESCRIPTION
**Pull request by JanCaha**
Expressions `to_base64` and `from_base64` including tests and docs.

### CODE REVIEW

File: tests/src/core/testqgsexpression.cpp [hunk 1]
```diff
@@ -1598,6 +1598,10 @@ class TestQgsExpression: public QObject
       QTest::newRow( "hash('QGIS', 'keccak_384')" ) << QStringLiteral( "hash('QGIS', 'keccak_384')" ) << false << QVariant( "c57a3aed9d856fa04e5eeee9b62b6e027cca81ba574116d3cc1f0d48a1ef9e5886ff463ea8d0fac772ee473bf92f810d" );
       QTest::newRow( "hash('QGIS', 'keccak_512')" ) << QStringLiteral( "hash('QGIS', 'keccak_512')" ) << false << QVariant( "6f0f751776b505e317de222508fa5d3ed7099d8f07c74fed54ccee6e7cdc6b89b4a085e309f2ee5210c942bbeb142bdfe48f84f912e0f3f41bdbf47110c2d344" );
 #endif
+      QTest::newRow( "to_base64 NULL" ) << QStringLiteral( "to_base64(NULL)" ) << false << QVariant();
+      QTest::newRow( "to_base64" ) << QStringLiteral( "to_base64('QGIS')" ) << false << QVariant( "UUdJUw==" );
+      QTest::newRow( "from_base64 NULL" ) << QStringLiteral( "from_base64(NULL)" ) << false << QVariant();
+      QTest::newRow( "from_base64" ) << QStringLiteral( "from_base64('UUdJUw==')" ) << false << QVariant( "QGIS" );
```
Comment (m-kuhn) [inline tests/src/core/testqgsexpression.cpp:1604]: Can you add some tests that work on a QByteArray?

File: src/core/expression/qgsexpressionfunction.cpp [hunk 2]
```diff
@@ -5545,6 +5545,25 @@ static QVariant fcnHashSha256( const QVariantList &values, const QgsExpressionCo
   return fcnHash( QgsExpressionUtils::getStringValue( values.at( 0 ), parent ), QCryptographicHash::Sha256 );
 }
 
+static QVariant fcnToBase64( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QString value = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
+
+  const QByteArray input = value.toLocal8Bit();
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionfunction.cpp:5552]: I think this should never go through a `QString` or a `toLocal8Bit` and be passed as raw bytes to tobase64.

File: src/core/expression/qgsexpressionfunction.cpp [hunk 3]
```diff
@@ -5545,6 +5545,25 @@ static QVariant fcnHashSha256( const QVariantList &values, const QgsExpressionCo
...
+}
+
+static QVariant fcnFromBase64( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QString value = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
+
+  const QByteArray base64 = value.toLocal8Bit();
+  const QByteArray decoded = QByteArray::fromBase64( base64 );
+
+  return QVariant( QString( decoded ) );
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionfunction.cpp:5564]: I don't think this should be forced to a QString here and preserved as raw bytes.

**Comment by m-kuhn** [comment]
Very nice functionality.
I wonder what your use case is though as you provide conversion between "strings" and "base64", whereas I would expect the conversion to be between "bytes" and "base64" to unleash the full power.

**Comment by JanCaha** [comment]
Hm, I see your point. Actually, this was suggested by @nyalldawson and if I understand him correctly, then this implementation follows his suggestion. So he is more qualified to provide the use case.

Maybe the `from_base64` function could take both string and QByteArray as an input? It is possible to have the encoded data stored in both these types.

File: src/core/expression/qgsexpressionfunction.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/expression/qgsexpressionfunction.cpp:5552]: Right - this should be:

    const QByteArray input = values.at( 0 ).toByteArray();
    return QVariant( QString( input.toBase64() ) );

File: src/core/expression/qgsexpressionfunction.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/core/expression/qgsexpressionfunction.cpp:5564]: And this one should be:

     const QString value = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
     const QByteArray base64 = value.toLocal8Bit();
     const QByteArray decoded = QByteArray::fromBase64( base64 );
     return QVariant( decoded );

**Comment by nyalldawson** [comment]
> Actually, this was suggested by @nyalldawson and if I understand him correctly, then this implementation follows his suggestion.

I think my email was just badly worded -- but the implementation here is flipped from what's needed.

**Comment by JanCaha** [comment]
I fixed the implementation using @nyalldawson suggestions.

Had to add a new case for `run_evaluation_test` function in `testqgsexpression.cpp` as it did not supported `QByteArray`. Hopefully, it will not break anything.

File: resources/function_help/json/from_base64 [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
+{
+  "name": "from_base64",
+  "type": "function",
+  "description": "Decodes a provided string in Base64 into readable string.",
```
Comment (nyalldawson) [inline resources/function_help/json/from_base64:5]:
```suggestion
  "description": "Decodes a provided string in Base64 into a binary value.",
```

File: resources/function_help/json/to_base64 [hunk 5]
```diff
@@ -0,0 +1,10 @@
+
+{
+  "name": "to_base64",
+  "type": "function",
+  "description": "Encodes a provided string using Base64.",
```
Comment (nyalldawson) [inline resources/function_help/json/to_base64:5]:
```suggestion
  "description": "Encodes a provided binary value into a string, using Base64 encoding.",
```

File: resources/function_help/json/to_base64 [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+  "arguments": [
+    {"arg":"string", "description": "the string to encode"}],
```
Comment (nyalldawson) [inline resources/function_help/json/to_base64:7]:
```suggestion
    {"arg":"string", "description": "the binary value to encode"}],
```

File: src/core/expression/qgsexpressionfunction.cpp [hunk 7]
```diff
@@ -5851,6 +5865,12 @@ const QList<QgsExpressionFunction *> &QgsExpression::Functions()
         << new QgsStaticExpressionFunction( QStringLiteral( "sha256" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "string" ) ),
                                             fcnHashSha256, QStringLiteral( "Conversions" ) )
 
+        //base64
+        << new QgsStaticExpressionFunction( QStringLiteral( "to_base64" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "string" ) ),
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp:5869]:
```suggestion
        << new QgsStaticExpressionFunction( QStringLiteral( "to_base64" ), QgsExpressionFunction::ParameterList() << QgsExpressionFunction::Parameter( QStringLiteral( "value" ) ),
```
(or "binary"?)

**Review by nyalldawson — COMMENTED** [review]
Nearly there! Just some documentation improvements to go

**Comment by nyalldawson** [comment]
Marking as freeze exempt given the prior approval and only doc wording left to address

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/36563

**Pull request closed without merging**